index.src.html

```diff
@@ -3415,29 +3417,39 @@
   VideoFrame(data, init)
 </dfn>
 1. If |init| is not a [=valid VideoFrameBufferInit=], throw a {{TypeError}}.
-2. Let |defaultRect| be «[ "x:" → `0`, "y" → `0`,
+1. Let |defaultRect| be «[ "x:" → `0`, "y" → `0`,
     "width" → |init|.{{VideoFrameBufferInit/codedWidth}}, "height" →
         |init|.{{VideoFrameBufferInit/codedWidth}} ]».
-3. Let |overrideRect| be `undefined`.
-4. If |init|.{{VideoFrameBufferInit/visibleRect}} [=map/exists=], assign its
+1. Let |overrideRect| be `undefined`.
+1. If |init|.{{VideoFrameBufferInit/visibleRect}} [=map/exists=], assign its
     value to |overrideRect|.
-4. Let |parsedRect| be the result of running the [=VideoFrame/Parse Visible
+1. Let |parsedRect| be the result of running the [=VideoFrame/Parse Visible
         Rect=] algorithm with |defaultRect|, |overrideRect|,
         |init|.{{VideoFrameBufferInit/codedWidth}},
         |init|.{{VideoFrameBufferInit/codedHeight}}, and
         |init|.{{VideoFrameBufferInit/format}}.
-5. If |parsedRect| is an exception, return |parsedRect|.
-6. Let |optLayout| be `undefined`.
-7. If |options|.{{VideoFrameBufferInit/layout}} [=map/exists=], assign its value
+1. If |parsedRect| is an exception, return |parsedRect|.
+1. Let |optLayout| be `undefined`.
+1. If |options|.{{VideoFrameBufferInit/layout}} [=map/exists=], assign its value
     to |optLayout|.
-8. Let |combinedLayout| be the result of running the [=VideoFrame/Compute
+1. Let |combinedLayout| be the result of running the [=VideoFrame/Compute
     Layout and Allocation Size=] algorithm with |parsedRect|,
     |init|.{{VideoFrameBufferInit/format}}, and |optLayout|.
-9. If |combinedLayout| is an exception, throw |combinedLayout|.
-10. If `data.byteLength` is less than |combinedLayout|’s
+1. If |combinedLayout| is an exception, throw |combinedLayout|.
+1. If `data.byteLength` is less than |combinedLayout|’s
     [=combined buffer layout/allocationSize=], throw a {{TypeError}}.
-11. Let |resource| be a new [=media resource=] containing a copy of |data|. Use
-        {{VideoFrameBufferInit/visibleRect}} and {{VideoFrameBufferInit/layout}}
+1. For each |transferable| in |init|.{{VideoFrameBufferInit/transfer}}:
+    1. If {{platform object/[[Detached]]}} internal slot is `true`,
+        then throw a {{DataCloneError}} {{DOMException}}.
+1. If |init|.{{VideoFrameBufferInit/transfer}} contains an ArrayBuffer referenced
+    by |data|:
+    1. Let |resource| be a new [=media resource=] referencing pixels in |data|
+        without a copy.
+        Use {{VideoFrameBufferInit/visibleRect}} and {{VideoFrameBufferInit/layout}}
+        to determine where in |data| the pixels for each plane reside.
```

Comment:
This is very hand-wavy, is there an algorithm to link to? Or maybe it just follows from the rest of the prose and can be omitted.

Reply:
I would prefer if implementations were allowed to copy. In general that's not observable here (unless we add buffer return), but while copying Chrome does try to minimize the coded size and that would be visible if it were not allowed on this path. It also allows us to shift the visible offset to the origin, which is more compatible; not all code is written to handle offsets (because they are rare in practice).

Chrome also expands the Y plane for odd-sized frames so that the coded size is always even. At some point in the future we may trust all of our processing paths enough to remove that, but for now we are likely to prefer to copy in that case.

Reply:
I think it's preferable to be explicit, since it's opt-in.

Reply:
I do not want to throw in edge cases like this, especially the odd-size one which I expect to go away in the future. I don't think developers should need to be aware that transfers can fail and always have to implement fallbacks; I can't imagine the fallback would ever be anything other than trying again without the transfer.

Reply:
I made transferring optional. Please take a look.